src/main/java/htsjdk/samtools/util/AsyncReadTaskRunner.java

```diff
@@ -0,0 +1,274 @@
+package htsjdk.samtools.util;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.*;
+
+/**
+ * Helper class for performing asynchronous reading.
+ * This class handles the asynchronous task scheduling.
+ * Asynchronous tasks are split into single blocking read ahead task (performReadAhead())
+ * and any number of non-blocking transform tasks (transform()).
+ * Synchronisation of the underlying stream/iterator is handled by ensuring that performReadAhead() is not executed
+ * by more than one thread concurrently.
+ * As this represents a serial bottleneck limiting palatalisation,
+ * implementations should ensure that as much computation work (such as decompression or record parsing) as possible
+ * is performed in transform().
+ */
+public abstract class AsyncReadTaskRunner<T, U> {
+    private static final Log log = Log.getInstance(AsyncReadTaskRunner.class);
+    private static Executor nonblockingThreadpool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors(), r -> {
+        Thread t = Executors.defaultThreadFactory().newThread(r);
+        t.setName("htsjdk-asyncio-nonblocking");
+        t.setDaemon(true);
+        return t;
+    });
+
+    private static Executor blockingThreadpool = Executors.newCachedThreadPool(r -> {

```

Comment:
Correct. This class ensures that we don't attempt to read from the underlying stream on multiplt threads at one.

This approach reduces the number of threads created to the number of read-ahead in flight at any one time. Generally speaking, this will be much lower count when you have many files open. A private thread per file introduces an overhead of over 1MB (default stack size) per file and that memory is better used in actual read-ahead buffers. We don't want to spin up OS threads unless they're actually going to be used.

The shared pool also allows the scheduling of these tasks to configured by the end user by replacing the executor.

Reply:
Unless you're pulling your bams from the internet, the bottleneck will be in the decompression and parsing. In my tracing, I've only ever got up to a handful of blocking threads even when processing 100+ files in parallel.